### ISSUE TITLE
Q: (possibly for FAQ): How are unknown fields handled

### ISSUE THREAD

**Issue by zellyn** [jviotti/jsonbinpack#162]
On of the originally important properties of protobufs was the way they handle unknown fields, and can pass them on to downstream apps without messing with them. JSON pretty much works the same if you just leave sub-branches alone. How does your format do with those?  (Important because schemas evolve, and you might see new data while still using the old schema.)

**Comment by jviotti** [comment]
Hey @zellyn ,

This is a very good question, and I'll make a note to explain on the FAQ. I'll write it up better later, but in summary:

JSON Schema is a constrain-based language. You can be as strict or loose on defining your data as you wish. With JSON Schema, you can even go on a hybrid approach: be very precise on some aspects of the JSON document while leaving certain parts open-ended.

 JSON BinPack takes advantage of this by statically analyzing the JSON Schema definition and deriving encoding rules based on the characteristics of every sub schema. If your sub-schema is truly open ended (i.e. the `{}` wildcard JSON Schema), then JSON BinPack will fallback to a "schema-less" approach where i.e. both the keys and the values are represented. See https://github.com/jviotti/jsonbinpack/blob/main/docs/encoder/any.markdown.

So in summary: JSON BinPack handles unknown fields gracefully given the flexibility of JSON Schema. If your schema definition is strict, more strict encoding rules will be derived and less data will be encoded. If a field is "unknown" (i.e. it belongs to an open-ended part of the schema), then it is still correctly encoded.

I hope that helps!

**Comment by jviotti** [comment]
In the benchmark (https://sourcemeta.github.io/json-size-benchmark/), I also encode every document using the wildcard JSON Schema definition that matches anything and call that the "schema-less" mode of JSON BinPack.

That mode of operation is comparable to MessagePack, CBOR, etc but still features a lot of the space-efficiency tricks from other parts of JSON BinPack. The benchmark shows JSON BinPack is pure schema-less mode being as space-efficient or more space-efficient than every schema-less alternative too.

**Comment by zellyn** [comment]
That makes sense. What happens if you see an unknown field (presumably added after you were compiled) in a completely strict region of the schema?

**Comment by jviotti** [comment]
As long as that unknown field successfully validates against your JSON Schema definition in some way, it will be handled correctly out of the box.

Otherwise, you can create a new version of your JSON Schema definition that does recognize the field and map it accordingly using any schema evolution tools already available in the JSON Schema ecosystem, such as Cambria (https://www.inkandswitch.com/cambria/).

Protocol Buffers is able to gracefully detect even the latter case through the use of the type/id field tags at the expense of limiting the amount of space-efficiency optimizations you can perform.

**Comment by jviotti** [comment]
I added a FAQ section covering this on the new website design which is now available at https://www.jsonbinpack.org!
